feat: added new field estimated_time under start and end dates section of project doctype

meshramsaroj opened the pull request:
**Task Link:** https://bloomstack.com/desk#Form/Task/TASK-2021-00714

**Description:**  Added new field estimated_time under START AND END DATES section of project doctype.
                     Updated estimated_time based on task's expected time
**GIF:**

![update_estimate_time_from_prject_And_task](https://user-images.githubusercontent.com/53329367/143529779-d4734bf5-d8dc-47a7-9d4b-5ef78c8681b1.gif)

Updated Gif:

![estimated_time_is_updated](https://user-images.githubusercontent.com/53329367/144363820-9aedf3a4-5a72-4b39-be9a-0de59c0f16a1.gif)

meshramsaroj committed:
Added new field estimated_time under START AND END DATES section of project doctype
Updated estimated_time based on task's expected time

neilLasrado commented on erpnext/projects/doctype/project/project.py:
```diff
@@ -234,6 +234,14 @@ def notify(self):
 			frappe.publish_realtime('show_notification_alert', message=notification_doc.get("subject"), after_commit=True, user=user)
 
+
+@frappe.whitelist()
```
You don't need to whitelist this function. re-write the function within self

meshramsaroj replied:
removed whitelist function

vishdha replied:
While accessing any server-side method via frappe.call(), you need to whitelist the server-side method using decorator @frappe.whitelist.

@meshramsaroj  if frappe.call from js then need to add frappe.whitelist

meshramsaroj committed:
fix: removed whitelist function

sahil28297 commented:
@neilLasrado this will be triggered only when the project is updated, is this fine or should there be a on_update hook in task or a daily hook that updates the project estimated time regularly?

sahil28297 commented:
> @neilLasrado this will be triggered only when the project is updated, is this fine or should there be a on_update hook in task or a daily hook that updates the project estimated time regularly?

@meshramsaroj add a on_update hook in task that updates all the project linked to it

meshramsaroj committed:
resolved merge conflicts

modified on_update function of project.py

added on_update in task.py to update project's estimated time.

meshramsaroj commented:
> > @neilLasrado this will be triggered only when the project is updated, is this fine or should there be a on_update hook in task or a daily hook that updates the project estimated time regularly?
> 
> @meshramsaroj add a on_update hook in task that updates all the project linked to it

@sahil28297 I've added on_update in task.py to update project's estimated time

Anurag810 commented on erpnext/projects/doctype/project/project.json:
```diff
@@ -40,7 +40,9 @@
   "actual_start_date",
   "actual_time",
   "column_break_20",
+  "estimated_time",
```
the `estimated_time` is twice in .json file.

meshramsaroj replied:
removed the repeated field from.json file

Anurag810 commented on erpnext/projects/doctype/project/project.json:
```diff
@@ -40,7 +40,9 @@
   "column_break_20",
+  "estimated_time",
   "actual_end_date",
+  "estimated_time",
```
the 'estimated_time' is twice in .json file.

Anurag810 commented on erpnext/projects/doctype/project/project.json:
```diff
@@ -518,7 +526,6 @@
    "write": 1
   }
  ],
- "quick_entry": 1,
```
enable quick_entry agian as it was there before.

meshramsaroj replied:
fixed

meshramsaroj committed:
fix: fixed changes as suggested

sahil28297 commented on erpnext/projects/doctype/project/project.py:
```diff
@@ -234,6 +234,14 @@ def notify(self):
 			frappe.publish_realtime('show_notification_alert', message=notification_doc.get("subject"), after_commit=True, user=user)
 
+
+def	update_estimated_time(project_name):
```
```suggestion
def update_estimated_time(project_name):
```

sahil28297 replied:
also frappe.enqueue this function

sahil28297 commented on erpnext/projects/doctype/task/task.py:
```diff
@@ -143,6 +144,7 @@ def on_update(self):
 		self.populate_depends_on()
 		self.notify()
 		self.assign_todo()
+		update_project_estimated_time(self)
```
move this to after_save hook so that the save doesn't slow down because of this

meshramsaroj replied:
used after_save method to update estimated_time in project doctype

sahil28297 commented on erpnext/projects/doctype/project/project.py:
```diff
@@ -212,7 +212,7 @@ def send_welcome_email(self):
 	def on_update(self):
 		self.notify()
-
+		update_estimated_time(self.name)
```
move this to after_save hook so that the save doesn't slow down because of this

meshramsaroj replied:
used after_save method to update estimated_time in project doctype

meshramsaroj committed:
used after_save method to update estimated_time in project doctype

vishdha commented on erpnext/projects/doctype/task/task.js:
```diff
@@ -40,6 +40,18 @@ frappe.ui.form.on("Task", {
 	},
 
+	after_save: function (frm) {
+		console.log(frm.doc);
```
@meshramsaroj  Remove Console.log

meshramsaroj replied:
removed console.log

sahil28297 commented:
@meshramsaroj please put an updated GIF as well after testing the new PR

meshramsaroj committed:
removed console log from code

meshramsaroj commented:
> @meshramsaroj please put an updated GIF as well after testing the new PR

@sahil28297 added updated GIF

meshramsaroj commented:
closing this PR and created new PR from base branch of oct-sprint-1
PR link: https://github.com/Bloomstack/erpnext/pull/1843
